src/code.cloudfoundry.org/silk/daemon/vtep/converger_test.go

```diff
@@ -2,437 +2,806 @@ package vtep_test
-					HardwareAddr: remoteMac,
-				},
-			))
-		})
+				localLease := controller.Lease{
+					UnderlayIP:          "10.10.0.4",
+					OverlaySubnet:       fmt.Sprintf("%s/32", localOverlayIP),
+					OverlayHardwareAddr: localMac.String(),
+				}
+
+				leases = []controller.Lease{
+					localLease,
+					remoteLeaseOnNetwork1,
+					remoteLeaseOnNetwork2,
+					remoteSingleIPLeaseOnNetwork1,
+					remoteSingleIPLeaseOnNetwork2,
+				}
+			})
 
-		It("does not log anything about non-routable leases", func() {
-			err := converger.Converge(leases)
-			Expect(err).NotTo(HaveOccurred())
+			It("adds routing rule for each non-single IP remote lease with a src", func() {
+				err := converger.Converge(leases)
+				Expect(err).NotTo(HaveOccurred())
+
+				// RouteReplace Call count is called once for each non-SingleIP lease
+				Expect(fakeNetlink.RouteReplaceCallCount()).To(Equal(2))
+
+				addedRoute := fakeNetlink.RouteReplaceArgsForCall(0)
+				destGW, destNet, _ := net.ParseCIDR(remoteLeaseOnNetwork1.OverlaySubnet)
+				Expect(addedRoute).To(Equal(&netlink.Route{
+					LinkIndex: 42,
+					Scope:     netlink.SCOPE_UNIVERSE,
+					Dst:       destNet, // 10.255.19.0/24
+					Gw:        destGW,  // 10.255.19.0
+					Flags:     unix.RTNH_F_ONLINK,
+					Src:       localOverlayLease.IP, // in the same overlay cidr, so it gets the local IP
+				}))
 
-			Expect(logger.Logs()).To(HaveLen(0))
+				_, network2cidr, _ := net.ParseCIDR("10.250.0.0/16")
+				addedRoute = fakeNetlink.RouteReplaceArgsForCall(1)
+				destGW, destNet, _ = net.ParseCIDR(remoteLeaseOnNetwork2.OverlaySubnet)
+				Expect(addedRoute).To(Equal(&netlink.Route{
+					LinkIndex: 42,
+					Scope:     netlink.SCOPE_UNIVERSE,
+					Dst:       destNet, // 10.250.50.0/24
+					Gw:        destGW,  // 10.250.50.0
+					Flags:     unix.RTNH_F_ONLINK,
+					Src:       network2cidr.IP, // on a different overlay idr, so it gets the first IP of the entire cidr
+				}))
+			})
 		})
 
-		Context("when the a remote lease is removed", func() {
-			var (
-				deletedNeighs []netlink.Neigh
-				oldDestMac    net.HardwareAddr
-			)
+		Context("when the local lease is non-SingleIP", func() {
 			BeforeEach(func() {
+				_, localOverlayLease, _ := net.ParseCIDR("10.255.32.0/24")
+				converger = &vtep.Converger{
+					OverlayNetwork: overlayNetworks,
+					LocalSubnet:    localOverlayLease,
+					LocalVTEP:      localVTEP,
+					NetlinkAdapter: fakeNetlink,
+					Logger:         logger,
+					IsSingleIP:     false,
+				}
+
+				localLease := controller.Lease{
+					UnderlayIP:          "10.10.0.4",
+					OverlaySubnet:       "10.255.32.0/24", // local subnet on network 1
+					OverlayHardwareAddr: localMac.String(),
+				}
+
+				leases = []controller.Lease{
+					localLease,
+					remoteLeaseOnNetwork1,
+					remoteLeaseOnNetwork2,
+					remoteSingleIPLeaseOnNetwork1,
+					remoteSingleIPLeaseOnNetwork2,
+				}
+			})
+
+			It("adds routing rule for each non-single IP remote lease", func() {
```

Comment:
is it worth testing that the local network doesnt have route/arp/fdb rules added for it?

Reply:
i think this is implicitly tested already. I test the call count for each item. And I test that each call count is about a remote lease. thus nothing is called for the local lease.